elasticapm/instrumentation/packages/asyncio/asyncpg.py

```diff
@@ -0,0 +1,50 @@
+#  BSD 3-Clause License
+#
+#  Copyright (c) 2019, Elasticsearch BV
+#  All rights reserved.
+#
+#  Redistribution and use in source and binary forms, with or without
+#  modification, are permitted provided that the following conditions are met:
+#
+#  * Redistributions of source code must retain the above copyright notice, this
+#    list of conditions and the following disclaimer.
+#
+#  * Redistributions in binary form must reproduce the above copyright notice,
+#    this list of conditions and the following disclaimer in the documentation
+#    and/or other materials provided with the distribution.
+#
+#  * Neither the name of the copyright holder nor the names of its
+#    contributors may be used to endorse or promote products derived from
+#    this software without specific prior written permission.
+#
+#  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+#  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+#  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+#  DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+#  FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+#  DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+#  SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+#  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+#  OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+#  OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+from elasticapm.contrib.asyncio.traces import async_capture_span
+from elasticapm.instrumentation.packages.asyncio.base import AsyncAbstractInstrumentedModule
+from elasticapm.instrumentation.packages.dbapi2 import extract_signature
+
+
+class AsyncPGInstrumentation(AsyncAbstractInstrumentedModule):
+
+    name = "asyncpg"
+
+    instrument_list = [("asyncpg.connection", "Connection._do_execute")]
```

Comment:
`Connection.execute` optionally calls the Cython `BaseProtocol.query` method if there are no arguments. I would suggest changing the instrumentation to the higher level `Connection.execute` and `Connection.executemany`

I believe this will cause the `test_Select_Sleep` test to fail as that query will not call `Connection._do_execute`

Reply:
Could you please explain why do you think it would be more beneficial to go to a higher level? How having calls to the Cython `BaseProtocol.query` method, if there are no arguments, would be helpful?

The idea behind my implementation was to go as low as possible to be able to register any queries to the database.

Reply:
`_do_execute` does not catch everything.

take a look at the implementation of `Connection.execute` for example: https://github.com/MagicStack/asyncpg/blob/master/asyncpg/connection.py#L294-L297

if we pass in a fully prepared query, for example `await connection.execute("select * from pg_sleep(0.2);")` we don't end up running `_do_execute` but rather skip immediately to the Cython implementation `BaseProtocol.query` which is not instrumentable by the apm agent

Reply:
thanks for explaining!